sdk/identity/azure-identity/src/main/java/com/azure/identity/implementation/IdentityClient.java

```diff
@@ -114,9 +129,93 @@
         }
         this.tenantId = tenantId;
         this.clientId = clientId;
+        this.clientSecret = clientSecret;
+        this.certificatePath = certificatePath;
+        this.certificatePassword = certificatePassword;
         this.options = options;
     }
 
+    private ConfidentialClientApplication getConfidentialClientApplication() {
+        if (confidentialClientApplication != null) {
+            return confidentialClientApplication;
+        } else if (clientId == null) {
+            throw logger.logExceptionAsError(new IllegalArgumentException(
+                "A non-null value for client ID must be provided for user authentication."));
+        } else {
+            String authorityUrl = options.getAuthorityHost().replaceAll("/+$", "") + "/" + tenantId;
+            IClientCredential credential;
+            if (clientSecret != null) {
+                credential = ClientCredentialFactory.createFromSecret(clientSecret);
+            } else if (certificatePath != null) {
+                try {
+                    if (certificatePassword == null) {
+                        byte[] pemCertificateBytes = Files.readAllBytes(Paths.get(certificatePath));
+                        credential = ClientCredentialFactory.createFromCertificate(
+                            CertificateUtil.privateKeyFromPem(pemCertificateBytes),
+                            CertificateUtil.publicKeyFromPem(pemCertificateBytes));
+                    } else {
+                        credential = ClientCredentialFactory.createFromCertificate(
+                            new FileInputStream(certificatePath), certificatePassword);
+                    }
+                } catch (CertificateException
+                    | UnrecoverableKeyException
+                    | NoSuchAlgorithmException
+                    | KeyStoreException
+                    | NoSuchProviderException
+                    | IOException e) {
+                    throw logger.logExceptionAsError(new RuntimeException(
+                        "Failed to parse the certificate for the credential: " + e.getMessage(), e));
+                }
+            } else {
+                throw logger.logExceptionAsError(
+                    new IllegalArgumentException("Must provide client secret or client certificate path"));
+            }
+            ConfidentialClientApplication.Builder applicationBuilder =
+                ConfidentialClientApplication.builder(clientId, credential);
+            try {
+                applicationBuilder = applicationBuilder.authority(authorityUrl);
+            } catch (MalformedURLException e) {
+                throw logger.logExceptionAsWarning(new IllegalStateException(e));
+            }
```

Comment:
Would this be better off above attempting to retrieve the credential as this will fail without requiring IO or handling security? Basically, this is a lighter exception to have happen first.

Reply:
Unfortunately the ConfidentialClientApplication.Builder requires the credential to be passed in the constructor of the builder, not as a builder setter. So the authority has to be set after the credential.